### PR TITLE
[KEYCLOAK-8043] Allow prompt=none query parameter to be propagated to…

### PR DESCRIPTION
**Pull request by sguilhen**
… default IdP

Link to Jira: https://issues.jboss.org/browse/KEYCLOAK-8043
Link to Documentation PR: https://github.com/keycloak/keycloak-documentation/pull/642

Following the discussion in Jira and Stian's comments there I've introduced a "Support prompt=none - On/Off" configuration for the OIDC providers (defaults to Off). When enabled, it signals that the provider can handle redirects that have the prompt=none query parameter.

The redirect with prompt=none is only performed if the request can be automatically directed to a default identity provider that supports the param. This is either accomplished by the "kc_idp_hint" param or by setting up a default provider for the realm. If a default provider can't be found we can't assume a provider and the standard "login_required" error is returned in the response.

The scenario for this redirect is as follows: Realm A has a default IdP B and the user is already authenticated in B but not yet in A. When an auth request with prompt=none is sent to A (for instance, using the javascript adapter with the check-sso param), the authorization endpoint in A verifies that the user is not yet authenticated but instead of returning the "login_required" error right away it first checks if a default provider can be determined and if this provider supports prompt=none it redirects the request to the provider. The IdP then verifies the user is already authenticated and as a result the account is re-established in realm A without errors.

If the user hasn't authenticated in the default IdP, then the "login_required" error is returned as expected.

### CODE REVIEW

**Comment by sguilhen** [comment]
@mposolda could you please review this when you find some time? I'm working on the docs to include explanations about the new config in the oidc providers and I'll link it here when I'm done.

File: services/src/main/java/org/keycloak/protocol/AuthorizationEndpointBase.java
Comment (sguilhen) [inline services/src/main/java/org/keycloak/protocol/AuthorizationEndpointBase.java]: This code kinda duplicates to some extent the logic from the IdentityProviderAuthenticator to determine the default provider. An alternative would be to work on the Response that is returned when we call authenticateOnly() and see if that response contains a redirect (302) to the broker. We could then parse the Location URI to extract the provider alias and use it to check if the provider supports prompt=none redirects. In other words, if we get a 302 and URL matches the /broker/idp_alias/login structure we know it was assembled by the IdentityProviderAuthenticator and we can extract the idp_alias from the URL and use it.

I've attempted this alternative way but the resulting code looked a bit hacky to be honest with the URL manipulation. It is, however, the most efficient way to obtain the default IdP alias without going through the realm config or re-checking the kc_idp_hint param, so if you feel this is better than the code above I can switch to it.

File: themes/src/main/resources/theme/base/admin/messages/admin-messages_en.properties
Comment (sguilhen) [inline themes/src/main/resources/theme/base/admin/messages/admin-messages_en.properties]: This is another topic I wanted a feedback on. To me "Support prompt=none" is a good name for the new configuration but I'm open to suggestions.

Comment (mposolda) [reply themes/src/main/resources/theme/base/admin/messages/admin-messages_en.properties]: Suggested in description

File: services/src/main/java/org/keycloak/protocol/AuthorizationEndpointBase.java
Comment (mposolda) [reply services/src/main/java/org/keycloak/protocol/AuthorizationEndpointBase.java]: +1 that this code duplicates the login in IdentityProviderAuthenticator. I suggest another alternative, which won't require AuthorizationEndpointBase to be aware of any IDP specific logic - see description for more details.

**Review by mposolda — COMMENTED** [review]
I am sorry for late reply and sorry for adding so much feedback for this - IMO some additional work is needed to support this properly.

- From the option and tooltip, I am not sure if it is clear what exactly does it mean? I suggest to rename option to "Forwards prompt=none from client" and tooltip to be something like "This is just used together with Identity Provider Authenticator. In case that client sends request with prompt=none and user is not yet authenticated, the error won't be directly returned to client, but the request with prompt=none will be forwarded to this identity provider" .

- SAML protocol also contains support for "prompt=none" equivalent to handle "Passive login". But I guess we don't need to add support to SAML Identity provider in this PR and add SAML support later if needed?

- As you pointed, the code in AuthorizationEndpointBase contains lots of duplications of IdentityProviderAuthenticator. IMO it will be good if AuthorizationEndpointBase is completely independent of any specific details of any concrete authenticator implementations - hence ideally it shouldn't be aware of any Identity brokering details at all, but just deal with any generic authentication SPI. I suggest something like this:
-- Introduce constant for new AuthenticationSession note. Something like FORWARDED_PASSIVE_LOGIN. Maybe add this constant to AuthenticationProcessor or AuthenticationManager (as those classes already contains constant for similar things related to authentication flow)
-- The logic for forwarding to IDP will be in the IdentityProviderAuthenticator instead of AuthorizationEndpointBase. The IdentityProviderAuthenticator will add the FORWARDED_PASSIVE_LOGIN note to AuthenticationSessionModel in case that it forwards to IDP and it figures that "prompt=none" was used by the client.
-- The AuthorizationEndpointBase just needs to check here https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/protocol/AuthorizationEndpointBase.java#L118 if there is note FORWARDED_PASSIVE_LOGIN in the AuthenticationSessionModel. If yes, then it won't throws the error, but instead it will just return challenge back to the client. 
In other words the FORWARDED_PASSIVE_LOGIN note can be used by other (custom) authenticators to signal that they support some way of handling passive login.

- There is possible case when client forwards prompt=none, which is then forwarded to IDP, and user is not yet logged in IDP. Then IDP will return response like "interaction_required" or "login_required" as mentioned in OIDC specs. In this case, Keycloak should send the error like PASSIVE_LOGIN_REQUIRED back to the client instead of display some error page. Will this work? It seems that there is no test for this scenario?

- Final (but very tricky) thing is, that there are lots of cases when KC requires interaction from user *after* successful login to IDP. For example there can be cases like this:
-- Client sends the "prompt=none"
-- This is forwarded to IDP and user is already logged in IDP. So IDP will just redirect back to KC
-- Now KC requires some interaction from user. For example user is not yet linked to KC user, so "First Broker Login" flow is executed and some authenticators wants to display page to the user (EG. IdpReviewProfileAuthenticator, IdpConfirmLinkAuthenticator or some others). The same applies for postBrokerLogin flow, actionRequired or consents page as all of those can be displayed by KC after redirection from IDP. However if original client sent the "prompt=none" then KC shouldn't display any pages at all and should always send the error PASSIVE_LOGIN_REQUIRED. I am quite sure that the PR doesn't handle those cases actually and KC page will be displayed to the user. At least I am no seeing any tests for any of those scenarios. I am not sure how much work will need to be done to handle those scenarios... Hopefully it is doable in some pretty way just on few places - if there is note FORWARDED_PASSIVE_LOGIN in the AuthenticationSession, the KC should send the error to the client instead of display the page.

**Comment by sguilhen** [comment]
> I am sorry for late reply and sorry for adding so much feedback for this - IMO some additional work is needed to support this properly.

That is fine, don't worry about it
> 
> * From the option and tooltip, I am not sure if it is clear what exactly does it mean? I suggest to rename option to "Forwards prompt=none from client" and tooltip to be something like "This is just used together with Identity Provider Authenticator. In case that client sends request with prompt=none and user is not yet authenticated, the error won't be directly returned to client, but the request with prompt=none will be forwarded to this identity provider" .

Yeah, I agree we could use better name for the option and better tooltip description. Just not sure about "Forwards prompt=none from client" though as that could imply that the IDP is capable of forwarding something when in fact it is at the receiving end of the forward containing prompt=none. Perhaps "Accepts prompt=none forward from client"? Also, the description should take the kc_idp_hint in consideration, no? Something like "This is just used together with Identity Provider Authenticator or when kc_idp_hint points to this identity provider.  In case that client sends request with prompt=none and user is not yet authenticated, the error won't be directly returned to client, but the request with prompt=none will be forwarded to this identity provider".
> * SAML protocol also contains support for "prompt=none" equivalent to handle "Passive login". But I guess we don't need to add support to SAML Identity provider in this PR and add SAML support later if needed?

Yes, SAML has something called isPassive that has the same meaning of the OIDC prompt=none. However, as the Jira is specific to OIDC I chose not to handle SAML IDP in this PR. As you've said, we can add support for SAML later if needed.
> * As you pointed, the code in AuthorizationEndpointBase contains lots of duplications of IdentityProviderAuthenticator. IMO it will be good if AuthorizationEndpointBase is completely independent of any specific details of any concrete authenticator implementations - hence ideally it shouldn't be aware of any Identity brokering details at all, but just deal with any generic authentication SPI. I suggest something like this:
>   -- Introduce constant for new AuthenticationSession note. Something like FORWARDED_PASSIVE_LOGIN. Maybe add this constant to AuthenticationProcessor or AuthenticationManager (as those classes already contains constant for similar things related to authentication flow)
>   -- The logic for forwarding to IDP will be in the IdentityProviderAuthenticator instead of AuthorizationEndpointBase. The IdentityProviderAuthenticator will add the FORWARDED_PASSIVE_LOGIN note to AuthenticationSessionModel in case that it forwards to IDP and it figures that "prompt=none" was used by the client.
>   -- The AuthorizationEndpointBase just needs to check here https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/protocol/AuthorizationEndpointBase.java#L118 if there is note FORWARDED_PASSIVE_LOGIN in the AuthenticationSessionModel. If yes, then it won't throws the error, but instead it will just return challenge back to the client.
>   In other words the FORWARDED_PASSIVE_LOGIN note can be used by other (custom) authenticators to signal that they support some way of handling passive login.

Ah ok, thanks for the suggestion. I will look into rearranging things this way.
> * There is possible case when client forwards prompt=none, which is then forwarded to IDP, and user is not yet logged in IDP. Then IDP will return response like "interaction_required" or "login_required" as mentioned in OIDC specs. In this case, Keycloak should send the error like PASSIVE_LOGIN_REQUIRED back to the client instead of display some error page. Will this work? It seems that there is no test for this scenario?

You are absolutely right about this. It hasn't been tested and if that happes (user not logged into IDP) the browser displays an error message saying the request couldn't be fulfilled (infinite loop of redirects). The IDP returns the error=login_required to the broker and the broker attempts to authenticate again using the prompt=none which leads to the IDP throwing the error back to the broker and so on. Need to investigate how to properly handle this though. Seems to me the logic in the IdentityBrokerService needs to be reworked a bit when the it gets a "login_required" error so that it sends the error back to the client. Not entirely sure how to do it yet though.
> * Final (but very tricky) thing is, that there are lots of cases when KC requires interaction from user _after_ successful login to IDP. For example there can be cases like this:
>   -- Client sends the "prompt=none"
>   -- This is forwarded to IDP and user is already logged in IDP. So IDP will just redirect back to KC
>   -- Now KC requires some interaction from user. For example user is not yet linked to KC user, so "First Broker Login" flow is executed and some authenticators wants to display page to the user (EG. IdpReviewProfileAuthenticator, IdpConfirmLinkAuthenticator or some others). The same applies for postBrokerLogin flow, actionRequired or consents page as all of those can be displayed by KC after redirection from IDP. However if original client sent the "prompt=none" then KC shouldn't display any pages at all and should always send the error PASSIVE_LOGIN_REQUIRED. I am quite sure that the PR doesn't handle those cases actually and KC page will be displayed to the user. At least I am no seeing any tests for any of those scenarios. I am not sure how much work will need to be done to handle those scenarios... Hopefully it is doable in some pretty way just on few places - if there is note FORWARDED_PASSIVE_LOGIN in the AuthenticationSession, the KC should send the error to the client instead of display the page.

I'm not sure I follow. If the user was logged in IDP, then the pages you've mentioned should be displayed irrespective of the prompt=none param, no? If the user is not logged in IDP, then we would get an error signaling the authentication failed and thus no further interaction should be required anyway. I mean, if the auth failed, you would never display the review profile page, for example, irrespective of the prompt=none param in the original auth request. I might be missing something here but I don't see why we should always send the error instead of displaying the pages.

**Comment by mposolda** [comment]
> > I am sorry for late reply and sorry for adding so much feedback for this - IMO some additional work is needed to support this properly.
> 
> That is fine, don't worry about it
> 
> > * From the option and tooltip, I am not sure if it is clear what exactly does it mean? I suggest to rename option to "Forwards prompt=none from client" and tooltip to be something like "This is just used together with Identity Provider Authenticator. In case that client sends request with prompt=none and user is not yet authenticated, the error won't be directly returned to client, but the request with prompt=none will be forwarded to this identity provider" .
> 
> Yeah, I agree we could use better name for the option and better tooltip description. Just not sure about "Forwards prompt=none from client" though as that could imply that the IDP is capable of forwarding something when in fact it is at the receiving end of the forward containing prompt=none. Perhaps "Accepts prompt=none forward from client"? Also, the description should take the kc_idp_hint in consideration, no? Something like "This is just used together with Identity Provider Authenticator or when kc_idp_hint points to this identity provider. In case that client sends request with prompt=none and user is not yet authenticated, the error won't be directly returned to client, but the request with prompt=none will be forwarded to this identity provider".
> 

+1 for "Accepts prompt=none forward from client" . +1 for the tooltip you mentioned. The kc_idp_hint is handled by the IdentityProviderAuthenticator as well, however I agree that mention in the tooltip won't hurt and is probably better :)

> > * SAML protocol also contains support for "prompt=none" equivalent to handle "Passive login". But I guess we don't need to add support to SAML Identity provider in this PR and add SAML support later if needed?
> 
> Yes, SAML has something called isPassive that has the same meaning of the OIDC prompt=none. However, as the Jira is specific to OIDC I chose not to handle SAML IDP in this PR. As you've said, we can add support for SAML later if needed.

+1

> 
> > * As you pointed, the code in AuthorizationEndpointBase contains lots of duplications of IdentityProviderAuthenticator. IMO it will be good if AuthorizationEndpointBase is completely independent of any specific details of any concrete authenticator implementations - hence ideally it shouldn't be aware of any Identity brokering details at all, but just deal with any generic authentication SPI. I suggest something like this:
> >   -- Introduce constant for new AuthenticationSession note. Something like FORWARDED_PASSIVE_LOGIN. Maybe add this constant to AuthenticationProcessor or AuthenticationManager (as those classes already contains constant for similar things related to authentication flow)
> >   -- The logic for forwarding to IDP will be in the IdentityProviderAuthenticator instead of AuthorizationEndpointBase. The IdentityProviderAuthenticator will add the FORWARDED_PASSIVE_LOGIN note to AuthenticationSessionModel in case that it forwards to IDP and it figures that "prompt=none" was used by the client.
> >   -- The AuthorizationEndpointBase just needs to check here https://github.com/keycloak/keycloak/blob/master/services/src/main/java/org/keycloak/protocol/AuthorizationEndpointBase.java#L118 if there is note FORWARDED_PASSIVE_LOGIN in the AuthenticationSessionModel. If yes, then it won't throws the error, but instead it will just return challenge back to the client.
> >   In other words the FORWARDED_PASSIVE_LOGIN note can be used by other (custom) authenticators to signal that they support some way of handling passive login.
> 
> Ah ok, thanks for the suggestion. I will look into rearranging things this way.

Cool, Thanks!

> 
> > * There is possible case when client forwards prompt=none, which is then forwarded to IDP, and user is not yet logged in IDP. Then IDP will return response like "interaction_required" or "login_required" as mentioned in OIDC specs. In this case, Keycloak should send the error like PASSIVE_LOGIN_REQUIRED back to the client instead of display some error page. Will this work? It seems that there is no test for this scenario?
> 
> You are absolutely right about this. It hasn't been tested and if that happens (user not logged into IDP) the browser displays an error message saying the request couldn't be fulfilled (infinite loop of redirects). The IDP returns the error=login_required to the broker and the broker attempts to authenticate again using the prompt=none which leads to the IDP throwing the error back to the broker and so on. Need to investigate how to properly handle this though. Seems to me the logic in the IdentityBrokerService needs to be reworked a bit when the it gets a "login_required" error so that it sends the error back to the client. Not entirely sure how to do it yet though.
> 

I am also not 100% sure. Hopefully it is possible to do it in single place somewhere in IdentityBrokerService or LoginActionsService and there is no need to handle this in 10 places in the individual identityProvider implementations :) I hope that when IDP returned error and there is a note FORWARDED_PASSIVE_LOGIN in authSession, the error can be returned to the client through something like "loginProtocol.sendError(PASSIVE_LOGIN_REQUIRED)" .

> > * Final (but very tricky) thing is, that there are lots of cases when KC requires interaction from user _after_ successful login to IDP. For example there can be cases like this:
> >   -- Client sends the "prompt=none"
> >   -- This is forwarded to IDP and user is already logged in IDP. So IDP will just redirect back to KC
> >   -- Now KC requires some interaction from user. For example user is not yet linked to KC user, so "First Broker Login" flow is executed and some authenticators wants to display page to the user (EG. IdpReviewProfileAuthenticator, IdpConfirmLinkAuthenticator or some others). The same applies for postBrokerLogin flow, actionRequired or consents page as all of those can be displayed by KC after redirection from IDP. However if original client sent the "prompt=none" then KC shouldn't display any pages at all and should always send the error PASSIVE_LOGIN_REQUIRED. I am quite sure that the PR doesn't handle those cases actually and KC page will be displayed to the user. At least I am no seeing any tests for any of those scenarios. I am not sure how much work will need to be done to handle those scenarios... Hopefully it is doable in some pretty way just on few places - if there is note FORWARDED_PASSIVE_LOGIN in the AuthenticationSession, the KC should send the error to the client instead of display the page.
> 
> I'm not sure I follow. If the user was logged in IDP, then the pages you've mentioned should be displayed irrespective of the prompt=none param, no? If the user is not logged in IDP, then we would get an error signaling the authentication failed and thus no further interaction should be required anyway. I mean, if the auth failed, you would never display the review profile page, for example, irrespective of the prompt=none param in the original auth request. I might be missing something here but I don't see why we should always send the error instead of displaying the pages.

Well, according to the specification for prompt=none https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest, both the KC and IdentityProvider should never display any pages to the user. In other words, KC should always send the error to the client if any interaction is needed. So if you have flow like this:

1) Client sent "prompt=none" to KC to check if user is already logged in KC without a need of user interaction
2) KC forwarded the request with "prompt=none" to the IDP
3) User was already logged in IDP and hence IDP redirected back to KC
4) There is no KC user linked to this IDP yet, so KC triggered "First Broker Login" flow and the IdpReviewProfileAuthenticator displayed the profile page "Hey, please update your profile" . This is not OK in my opinion as KC page is displayed to the user, which is a break of "prompt=none" contract expected by the original client (from step 1).

There are lots of variants of step 4 - page could be displayed by any authenticator from "FirstBrokerLogin" flow or "PostBrokerLogin" flow. There can be also page displayed by the actionsRequired or consent page. In all those cases, should be error sent back to the client instead of KC displaying any page.

As in previous part (IDP returning error), I hope that this can be handled just in few places and not at the level of individual authenticators or requiredActions. ATM I am not sure if it is possible, maybe some more refactoring could be needed to this.

It is possible that forwarding "prompt=none" to IDP was the easy part and ensure that KC never display the page would be more challenging. But hopefully not... ;)

**Comment by sguilhen** [comment]
> +1 for "Accepts prompt=none forward from client" . +1 for the tooltip you mentioned. The kc_idp_hint is handled by the IdentityProviderAuthenticator as well, however I agree that mention in the tooltip won't hurt and is probably better :)

Changed the option and tooltip exactly as discussed above and also changed the code so that it uses and auth note as you suggested above. It is a lot cleaner and the note was a key piece in the implementation of the other issues you mentioned that required proper error messages to be redirected back to the client.
 
> I am also not 100% sure. Hopefully it is possible to do it in single place somewhere in IdentityBrokerService or LoginActionsService and there is no need to handle this in 10 places in the individual identityProvider implementations :) I hope that when IDP returned error and there is a note FORWARDED_PASSIVE_LOGIN in authSession, the error can be returned to the client through something like "loginProtocol.sendError(PASSIVE_LOGIN_REQUIRED)" .

Yeah, that was exactly what I did to solve this. The IdentityBrokerService now detects passive login errors that might have happened in the IDP and uses a LoginProtocol to send the error back to the client.

> Well, according to the specification for prompt=none https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest, both the KC and IdentityProvider should never display any pages to the user. In other words, KC should always send the error to the client if any interaction is needed. So if you have flow like this:
> 
> 1. Client sent "prompt=none" to KC to check if user is already logged in KC without a need of user interaction
> 2. KC forwarded the request with "prompt=none" to the IDP
> 3. User was already logged in IDP and hence IDP redirected back to KC
> 4. There is no KC user linked to this IDP yet, so KC triggered "First Broker Login" flow and the IdpReviewProfileAuthenticator displayed the profile page "Hey, please update your profile" . This is not OK in my opinion as KC page is displayed to the user, which is a break of "prompt=none" contract expected by the original client (from step 1).
> 
> There are lots of variants of step 4 - page could be displayed by any authenticator from "FirstBrokerLogin" flow or "PostBrokerLogin" flow. There can be also page displayed by the actionsRequired or consent page. In all those cases, should be error sent back to the client instead of KC displaying any page.
> 
> As in previous part (IDP returning error), I hope that this can be handled just in few places and not at the level of individual authenticators or requiredActions. ATM I am not sure if it is possible, maybe some more refactoring could be needed to this.
> 
> It is possible that forwarding "prompt=none" to IDP was the easy part and ensure that KC never display the page would be more challenging. But hopefully not... ;)

That is a very good point and after reading the spec again I totally agree with you. Using the note set by the IdentityProviderAuthenticator I was able to detect when any of the broker flows returns a response that would lead to a page requiring interaction from the user and instead return an interaction_required error (this was done in the LoginActionsService that handle both broker executions). Same was done to required actions and consents which the IdentityProviderService executes after the broker flows finish successfully.

I've also enhanced the test to check for all these scenarios: passive login errors thrown by the IDP, required actions and flows redirecting the other pages. Let me know what you think.

**Comment by sguilhen** [comment]
PR for documentation has been updated as well to reflect new name for the config option and it also includes a paragraph on errors that might be returned to the client if user is not authenticated in the IDP or if any screens requiring user interaction would be otherwise displayed.

**Review by mposolda — APPROVED** [review]
@sguilhen Thanks Stefan! IMO This is good to go when we have docs PR approval from Andrew.

**Comment by sguilhen** [comment]
@mposolda thanks for reviewing it!

**Pull request merged**
